mgmtfn/mesosplugin/cnidriver.go

```diff
@@ -0,0 +1,376 @@
+		nsCmds = append(nsCmds, gwCmd)
+		cniReq.cniSuccessResp.IP4.Gateway = nwState.Gateway
+		cniLog.Infof("ipv4 default gateway of endpoint %s", nwState.Gateway)
+	}
+
+	// ipv6
+	if len(mResp.EndpointConfig.IPv6Address) > 0 {
+		ipv6Cmd := []string{"ip", "-6", "addr", "add",
+			fmt.Sprintf("%s/%d", mResp.EndpointConfig.IPv6Address, nwState.IPv6SubnetLen),
+			"dev", cniReq.pluginArgs.CniIfname}
+		nsCmds = append(nsCmds, ipv6Cmd)
+		cniReq.cniSuccessResp.IP6.IPAddress = fmt.Sprintf("%s/%d", mResp.EndpointConfig.IPv6Address, nwState.IPv6SubnetLen)
+		cniLog.Infof("ipv6 address of endpoint %s", mResp.EndpointConfig.IPv6Address)
+
+	}
+
+	if len(nwState.IPv6Gateway) > 0 {
+		ipv6gwCmd := []string{"ip", "-6", "route", "add", "default", "via",
+			fmt.Sprintf("%s", nwState.IPv6Gateway)}
+		nsCmds = append(nsCmds, ipv6gwCmd)
+		cniReq.cniSuccessResp.IP6.Gateway = nwState.IPv6Gateway
+		cniLog.Infof("ipv6 gateway of endpoint %s", nwState.IPv6Gateway)
+	}
+
+	if len(nwState.DNSServer) > 0 {
+		cniReq.cniSuccessResp.Dns.NameServers = append(cniReq.cniSuccessResp.Dns.NameServers, nwState.DNSServer)
+	}
+
+	if _, err := cniReq.ipnsBatchExecute(cniReq.pluginArgs.CniContainerid, nsCmds); err != nil {
+		cniLog.Errorf("failed to execute commands in namespace %s: %s",
+			cniReq.pluginArgs.CniNetns, err.Error())
+		cniReq.unlinkNetNs()
+		return err
+	}
+
+	return nil
+}
+
+func (cniReq *cniServer) unlinkNetNs() error {
+	if err := os.Remove(netNsDir + cniReq.pluginArgs.CniContainerid); err != nil {
+		cniLog.Errorf("failed to unlink namespace %s, %s",
+			cniReq.pluginArgs.CniNetns, err)
+		return err
+	}
+	return nil
+}
+
+func (cniReq *cniServer) createCniEndPoint() error {
+	var err error
+	ovsEpDriver := &drivers.OvsOperEndpointState{}
+	ovsEpDriver.StateDriver = stateDriver
+
+	nwState := &mastercfg.CfgNetworkState{}
+	nwState.StateDriver = stateDriver
+
+	// build endpoint request
+	epReq := master.CreateEndpointRequest{
+		TenantName:  cniReq.endPointLabels[cniapi.LABEL_TENANT_NAME],
+		NetworkName: cniReq.endPointLabels[cniapi.LABEL_NETWORK_NAME],
+		ServiceName: cniReq.endPointLabels[cniapi.LABEL_NETWORK_GROUP],
+		EndpointID:  cniReq.pluginArgs.CniContainerid,
+		ConfigEP: intent.ConfigEP{
+			Container:   cniReq.pluginArgs.CniContainerid,
+			Host:        hostName,
+			ServiceName: cniReq.endPointLabels[cniapi.LABEL_NETWORK_GROUP],
+		},
+	}
+
+	cniLog.Infof("endpoint-req: epid:%s cont-id:%s ", epReq.EndpointID, epReq.ConfigEP.Container)
+
+	epResp := master.CreateEndpointResponse{}
+	if err = cluster.MasterPostReq("/plugin/createEndpoint", &epReq, &epResp); err != nil {
+		cniLog.Errorf("failed to create endpoint in master: %s", err.Error())
+		return err
+	}
+
+	cniLog.Infof("endpoint created  %+v", epResp)
+
+	// create endpoint in netplugin
+	if err = netPlugin.CreateEndpoint(cniReq.endpointID); err != nil {
+		cniLog.Errorf("failed to create endpoint in netplugin: %s", err.Error())
+		goto cleanupMaster
```

Comment:
we have to use multiple defer() or function calls to replace,
I think the code will be more readable with "goto" here.